presto-parser/src/main/java/com/facebook/presto/sql/tree/LogicalBinaryExpression.java

```diff
@@ -82,6 +82,24 @@ public Expression getRight()
         return right;
     }
 
+    public boolean needsJoinWarning()
+    {
+        if (this.operator == Operator.OR) {
```

Comment:
Consider implementing this with a `ArrayDequeue<Expression>` to remove the recursion.